[gofish] how to set BasicAuth for redfish provider

Hi, I encountered 401 error while using redfish provider, adding `BasicAuth: true` to config solved the issue, is there any way to pass the config down to gofish? thanks

```go
config := gofish.ClientConfig{
		Endpoint:   host,
		Username:   user,
		Password:   pass,
		Insecure:   true,
		HTTPClient: httpClient,
		BasicAuth:  true,
	}
```

```yml
Unauthorized
--- FAIL: TestBMC (0.57s)
    bmc_test.go:29: 
                Error Trace:    /home/github.com/-vm/pkg/bmc/bmc_test.go:29
                Error:          Not equal: 
                                expected: <nil>(<nil>)
                                actual  : *multierror.Error(*multierror.Error{Errors:[]error{provider: gofish: failed to retrieve some items: [{"link":"/redfish/v1/Systems/system","error":"401: Unauthorized"}], failed to get power state}, ErrorFormat:(multierror.ErrorFormatFunc)(nil)})
                Test:           TestBMC
                Messages:       BMC poweron https://10.2.7.5: failed: 2 errors occurred:
                                        * provider: gofish: failed to retrieve some items: [{"link":"/redfish/v1/Systems/system","error":"401: Unauthorized"}]
                                        * failed to get power state
```

to add more info, below is the log when I use `bmclib` w/o `BasicAuth` in gofish client config, is there any solution to solve the issue other than passing down the  `BasicAuth` to gofish, thanks.
 
```yml
GET /redfish/v1/ HTTP/1.1
Host: 10.22.7.5
User-Agent: gofish/1.0
Connection: close
Accept: application/json
Content-Type: application/json
Accept-Encoding: gzip


HTTP/1.1 200 OK
Connection: close
Content-Length: 1303
Cache-Control: no-Store,no-Cache
Content-Security-Policy: default-src 'none'; img-src 'self' data:; font-src 'self'; style-src 'self'; script-src 'self'; connect-src 'self' wss:; form-action 'none'; frame-ancestors 'none'; plugin-types 'none'; base-uri 'none'
Content-Type: application/json
Date: Wed, 03 May 2023 13:53:04 GMT
Pragma: no-cache
Strict-Transport-Security: max-age=31536000; includeSubdomains; preload
X-Content-Type-Options: nosniff
X-Frame-Options: DENY
X-Xss-Protection: 1; mode=block

{
  "@odata.id": "/redfish/v1",
  "@odata.type": "#ServiceRoot.v1_5_0.ServiceRoot",
  "AccountService": {
    "@odata.id": "/redfish/v1/AccountService"
  },
  "CertificateService": {
    "@odata.id": "/redfish/v1/CertificateService"
  },
  "Chassis": {
    "@odata.id": "/redfish/v1/Chassis"
  },
  "EventService": {
    "@odata.id": "/redfish/v1/EventService"
  },
  "Id": "RootService",
  "JsonSchemas": {
    "@odata.id": "/redfish/v1/JsonSchemas"
  },
  "Links": {
    "Sessions": {
      "@odata.id": "/redfish/v1/SessionService/Sessions"
    }
  },
  "Managers": {
    "@odata.id": "/redfish/v1/Managers"
  },
  "Name": "Root Service",
  "Oem": {
    "OpenBmc": {
      "@odata.type": "#OemServiceRoot.v1_0_0.ServiceRoot",
      "CupsService": {
        "@odata.id": "/redfish/v1/CupsService"
      }
    }
  },
  "RedfishVersion": "1.9.0",
  "Registries": {
    "@odata.id": "/redfish/v1/Registries"
  },
  "SessionService": {
    "@odata.id": "/redfish/v1/SessionService"
  },
  "Systems": {
    "@odata.id": "/redfish/v1/Systems"
  },
  "Tasks": {
    "@odata.id": "/redfish/v1/TaskService"
  },
  "TelemetryService": {
    "@odata.id": "/redfish/v1/TelemetryService"
  },
  "UUID": "87c34931-c0d1-54db-a5c9-6b97cd6d049f",
  "UpdateService": {
    "@odata.id": "/redfish/v1/UpdateService"
  }
}
POST /redfish/v1/SessionService/Sessions HTTP/1.1
Host: 10.22.7.5
User-Agent: gofish/1.0
Connection: close
Content-Length: 45
Accept: application/json
Content-Type: application/json
Accept-Encoding: gzip

{"UserName":"user","Password":"xxx"}
HTTP/1.1 201 Created
Connection: close
Content-Length: 268
Cache-Control: no-Store,no-Cache
Content-Security-Policy: default-src 'none'; img-src 'self' data:; font-src 'self'; style-src 'self'; script-src 'self'; connect-src 'self' wss:; form-action 'none'; frame-ancestors 'none'; plugin-types 'none'; base-uri 'none'
Content-Type: application/json
Date: Wed, 03 May 2023 13:53:04 GMT
Location: /redfish/v1/SessionService/Sessions/VF8KSqNWI1
Pragma: no-cache
Strict-Transport-Security: max-age=31536000; includeSubdomains; preload
X-Auth-Token: 7ix2wlLARkRa253ZDXlX
X-Content-Type-Options: nosniff
X-Frame-Options: DENY
X-Xss-Protection: 1; mode=block

{
  "@odata.id": "/redfish/v1/SessionService/Sessions/VF8KSqNWI1",
  "@odata.type": "#Session.v1_3_0.Session",
  "ClientOriginIPAddress": "10.226.76.8",
  "Description": "Manager User Session",
  "Id": "VF8KSqNWI1",
  "Name": "User Session",
  "UserName": "user"
}
GET /redfish/v1/Systems HTTP/1.1
Host: 10.22.7.5
User-Agent: gofish/1.0
Connection: close
Accept: application/json
Content-Type: application/json
Cookie: sessionKey=7ix2wlLARkRa253ZDXlX
X-Auth-Token: 7ix2wlLARkRa253ZDXlX
Accept-Encoding: gzip


HTTP/1.1 200 OK
Connection: close
Content-Length: 259
Cache-Control: no-Store,no-Cache
Content-Security-Policy: default-src 'none'; img-src 'self' data:; font-src 'self'; style-src 'self'; script-src 'self'; connect-src 'self' wss:; form-action 'none'; frame-ancestors 'none'; plugin-types 'none'; base-uri 'none'
Content-Type: application/json
Date: Wed, 03 May 2023 13:53:04 GMT
Pragma: no-cache
Strict-Transport-Security: max-age=31536000; includeSubdomains; preload
X-Content-Type-Options: nosniff
X-Frame-Options: DENY
X-Xss-Protection: 1; mode=block

{
  "@odata.id": "/redfish/v1/Systems",
  "@odata.type": "#ComputerSystemCollection.ComputerSystemCollection",
  "Members": [
    {
      "@odata.id": "/redfish/v1/Systems/system"
    }
  ],
  "Members@odata.count": 1,
  "Name": "Computer System Collection"
}
GET /redfish/v1/Systems/system HTTP/1.1
Host: 10.22.7.5
User-Agent: gofish/1.0
Connection: close
Accept: application/json
Content-Type: application/json
Cookie: sessionKey=7ix2wlLARkRa253ZDXlX
X-Auth-Token: 7ix2wlLARkRa253ZDXlX
Accept-Encoding: gzip


HTTP/1.1 401 Unauthorized
Connection: close
Content-Length: 12
Cache-Control: no-Store,no-Cache
Content-Security-Policy: default-src 'none'; img-src 'self' data:; font-src 'self'; style-src 'self'; script-src 'self'; connect-src 'self' wss:; form-action 'none'; frame-ancestors 'none'; plugin-types 'none'; base-uri 'none'
Date: Wed, 03 May 2023 13:53:04 GMT
Pragma: no-cache
Strict-Transport-Security: max-age=31536000; includeSubdomains; preload
X-Content-Type-Options: nosniff
X-Frame-Options: DENY
X-Xss-Protection: 1; mode=block

Unauthorized
DELETE /redfish/v1/SessionService/Sessions/VF8KSqNWI1 HTTP/1.1
Host: 10.22.7.5
User-Agent: gofish/1.0
Connection: close
Accept: application/json
Content-Type: application/json
Cookie: sessionKey=7ix2wlLARkRa253ZDXlX
X-Auth-Token: 7ix2wlLARkRa253ZDXlX
Accept-Encoding: gzip


HTTP/1.1 401 Unauthorized
Connection: close
Content-Length: 12
Cache-Control: no-Store,no-Cache
Content-Security-Policy: default-src 'none'; img-src 'self' data:; font-src 'self'; style-src 'self'; script-src 'self'; connect-src 'self' wss:; form-action 'none'; frame-ancestors 'none'; plugin-types 'none'; base-uri 'none'
Date: Wed, 03 May 2023 13:53:04 GMT
Pragma: no-cache
Strict-Transport-Security: max-age=31536000; includeSubdomains; preload
X-Content-Type-Options: nosniff
X-Frame-Options: DENY
X-Xss-Protection: 1; mode=block

Unauthorized
--- FAIL: TestBMC (0.57s)
    bmc_test.go:29: 
                Error Trace:    /home/github.com/vm/pkg/bmc/bmc_test.go:29
                Error:          Not equal: 
                                expected: <nil>(<nil>)
                                actual  : *multierror.Error(*multierror.Error{Errors:[]error{provider: gofish: failed to retrieve some items: [{"link":"/redfish/v1/Systems/system","error":"401: Unauthorized"}], failed to get power state}, ErrorFormat:(multierror.ErrorFormatFunc)(nil)})
                Test:           TestBMC
                Messages:       BMC poweron https://10.22.7.5: failed: 2 errors occurred:
                                        * provider: gofish: failed to retrieve some items: [{"link":"/redfish/v1/Systems/system","error":"401: Unauthorized"}]
                                        * failed to get power state
```

--- a/client.go
+++ b/client.go
@@ -34,6 +34,7 @@ type Client struct {
 	metadata                     *bmc.Metadata
 	perProviderTimeout           func(context.Context) time.Duration
 	redfishVersionsNotCompatible []string
+	redfishBasicAuthEnabled      bool
 }
 
 const (
@@ -100,6 +101,13 @@ func WithRedfishVersionsNotCompatible(versions []string) Option {
 	}
 }
 
+// WithRedfishBasicAuth enables Basic Authentication on the Redfish driver.
+func WithRedfishBasicAuth() Option {
+	return func(args *Client) {
+		args.redfishBasicAuthEnabled = true
+	}
+}
+
 // NewClient returns a new Client struct
 func NewClient(host, port, user, pass string, opts ...Option) *Client {
 	var defaultClient = &Client{
@@ -159,7 +167,12 @@ func (c *Client) registerProviders() {
 	c.Registry.Register(asrockrack.ProviderName, asrockrack.ProviderProtocol, asrockrack.Features, nil, driverAsrockrack)
 
 	// register gofish provider
-	driverGoFish := redfish.New(c.Auth.Host, c.Auth.Port, c.Auth.User, c.Auth.Pass, c.Logger, redfishwrapper.WithHTTPClient(c.httpClient), redfishwrapper.WithVersionsNotCompatible(c.redfishVersionsNotCompatible))
+	redfishOpts := []redfishwrapper.Option{
+		redfishwrapper.WithHTTPClient(c.httpClient),
+		redfishwrapper.WithVersionsNotCompatible(c.redfishVersionsNotCompatible),
+		redfishwrapper.WithBasicAuthEnabled(c.redfishBasicAuthEnabled),
+	}
+	driverGoFish := redfish.New(c.Auth.Host, c.Auth.Port, c.Auth.User, c.Auth.Pass, c.Logger, redfishOpts...)
 	c.Registry.Register(redfish.ProviderName, redfish.ProviderProtocol, redfish.Features, nil, driverGoFish)
 
 	// register AMT provider

--- a/client_test.go
+++ b/client_test.go
@@ -83,6 +83,38 @@ func TestWithRedfishVersionsNotCompatible(t *testing.T) {
 	}
 }
 
+func TestWithRedfishBasicAuth(t *testing.T) {
+	host := "127.0.0.1"
+	port := "623"
+	user := "ADMIN"
+	pass := "ADMIN"
+
+	tests := []struct {
+		name    string
+		enabled bool
+	}{
+		{
+			"disabled",
+			false,
+		},
+		{
+			"enabled",
+			true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			var opts []Option
+			if tt.enabled {
+				opts = append(opts, WithRedfishBasicAuth())
+			}
+
+			cl := NewClient(host, port, user, pass, opts...)
+			assert.Equal(t, tt.enabled, cl.redfishBasicAuthEnabled)
+		})
+	}
+}
+
 func TestWithConnectionTimeout(t *testing.T) {
 	host := "127.0.0.1"
 	port := "623"

--- a/examples/status/main.go
+++ b/examples/status/main.go
@@ -30,7 +30,10 @@ func main() {
 		l.Fatal("required host/user/pass parameters not defined")
 	}
 
-	clientOpts := []bmclib.Option{bmclib.WithLogger(logger)}
+	clientOpts := []bmclib.Option{
+		bmclib.WithLogger(logger),
+		bmclib.WithRedfishBasicAuth(),
+	}
 
 	if *withSecureTLS {
 		var pool *x509.CertPool
@@ -58,18 +61,9 @@ func main() {
 	}
 	defer cl.Close(ctx)
 
-	inventory, err := cl.Inventory(ctx)
-	if err != nil {
-		l.Fatal(err)
-	}
-
-	l.WithField("bmc-version", inventory.BMC.Firmware.Installed).Info()
-
 	state, err := cl.GetPowerState(ctx)
 	if err != nil {
 		l.WithError(err).Error()
 	}
 	l.WithField("power-state", state).Info()
-
-	l.WithField("bios-version", inventory.BIOS.Firmware.Installed).Info()
 }

--- a/internal/redfishwrapper/client.go
+++ b/internal/redfishwrapper/client.go
@@ -22,6 +22,7 @@ type Client struct {
 	port                  string
 	user                  string
 	pass                  string
+	basicAuth             bool
 	versionsNotCompatible []string // a slice of redfish versions to ignore as incompatible
 	client                *gofish.APIClient
 	httpClient            *http.Client
@@ -55,6 +56,13 @@ func WithVersionsNotCompatible(versions []string) Option {
 	}
 }
 
+// WithBasicAuthEnabled sets Basic Auth on the Gofish driver.
+func WithBasicAuthEnabled(e bool) Option {
+	return func(c *Client) {
+		c.basicAuth = e
+	}
+}
+
 // NewClient returns a redfishwrapper client
 func NewClient(host, port, user, pass string, opts ...Option) *Client {
 	if !strings.HasPrefix(host, "https://") && !strings.HasPrefix(host, "http://") {
@@ -84,6 +92,7 @@ func (c *Client) Open(ctx context.Context) error {
 		Password:   c.pass,
 		Insecure:   true,
 		HTTPClient: c.httpClient,
+		BasicAuth:  c.basicAuth,
 	}
 
 	if config.HTTPClient == nil {
@@ -136,6 +145,11 @@ func (c *Client) Close(ctx context.Context) error {
 func (c *Client) SessionActive() error {
 	if c.client == nil || c.client.Service == nil {
 		return bmclibErrs.ErrNotAuthenticated
+	}
+
+	// With basic auth enabled theres no session to be checked.
+	if c.basicAuth {
+		return nil
 	}
 
 	_, err := c.client.GetSession()

--- a/internal/redfishwrapper/client_test.go
+++ b/internal/redfishwrapper/client_test.go
@@ -31,3 +31,29 @@ func TestWithVersionsNotCompatible(t *testing.T) {
 		})
 	}
 }
+
+func TestWithBasicAuthEnabled(t *testing.T) {
+	host := "127.0.0.1"
+	user := "ADMIN"
+	pass := "ADMIN"
+
+	tests := []struct {
+		name    string
+		enabled bool
+	}{
+		{
+			"disabled",
+			false,
+		},
+		{
+			"enabled",
+			true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			client := NewClient(host, "", user, pass, WithBasicAuthEnabled(tt.enabled))
+			assert.Equal(t, tt.enabled, client.basicAuth)
+		})
+	}
+}

